tests/transforms/test_transform_can_be_hoisted.py

```diff
@@ -0,0 +1,138 @@
+from xdsl.builder import Builder
+from xdsl.dialects import builtin, func, scf, test
+from xdsl.dialects.arith import Addf, Addi, Constant, DivUI, Muli
+from xdsl.dialects.builtin import (
+    FloatAttr,
+    FunctionType,
+    IndexType,
+    IntegerAttr,
+    ModuleOp,
+    f32,
+    i32,
+)
+from xdsl.dialects.scf import For, Yield
+from xdsl.ir import Block, BlockArgument, Region
+from xdsl.transforms.loop_invariant_code_motion import can_be_hoisted
+
+index = IndexType()
+
+
+def test_is_loop_dependent_no_dep():
+    lb = Constant.from_int_and_width(0, i32)
+    ub = Constant.from_int_and_width(42, i32)
+    step = Constant.from_int_and_width(3, i32)
+
+    op1 = test.TestOp(result_types=[i32])
+    op2 = test.TestOp(result_types=[i32])
+
+    for_op = For(lb, ub, step, [], Block([op1, op2], arg_types=[i32]))
+
+    assert can_be_hoisted(op1, for_op.body)
+    assert can_be_hoisted(op2, for_op.body)
+
+
+def test_for_with_loop_invariant_verify():
+    """Test for with loop-variant variables"""
+
+    lower = Constant.from_int_and_width(0, IndexType())
+    upper = Constant.from_int_and_width(42, IndexType())
+    step = Constant.from_int_and_width(3, IndexType())
+    carried = Constant.from_int_and_width(1, IndexType())
+
+    @Builder.implicit_region((IndexType(), IndexType()))
+    def body(_: tuple[BlockArgument, ...]) -> None:
+        a = Constant(IntegerAttr.from_int_and_width(1, 32), i32)
+        b = Constant(IntegerAttr.from_int_and_width(2, 32), i32)
+        # Operations on these constants
+        c = Addi(a, b)
+        d = Muli(a, b)
+        e = DivUI(c, d)
+        Yield(e)
+
+    for_op = For(lower, upper, step, [carried], body)
+
+    for op in for_op.body.walk():
+        for regoin in for_op.regions:
+            # MLIR iterates the for loop's body and hoists operations that
+            # can_be_hoisted. However, the DivUI operation cannot be hoisted immediately
+            # because its operands are still within the loop's body. It can only be
+            # hoisted once the instructions generating those operands have been hoisted.
+            if not (isinstance(op, Yield) or isinstance(op, DivUI)):
+                assert can_be_hoisted(op, regoin)
```

Comment:
Functions encapsulate variables, but we want to be able to test them directly, so the other way of building blocks is probably preferable here:
```suggestion
    block = Block(arg_types=(index, index,))
    region = Region(block)
    with ImplicitBuilder(block) as i_arg, carried_arg:
        a = Constant(IntegerAttr.from_int_and_width(1, 32), i32)
        b = Constant(IntegerAttr.from_int_and_width(2, 32), i32)
        # Operations on these constants
        c = Addi(a, b)
        d = Muli(a, b)
        e = DivUI(c, d)
        f = Yield(e)

    for_op = For(lower, upper, step, [carried], region)

    assert can_be_hoisted(region, a)
    assert can_be_hoisted(region, b)
    assert can_be_hoisted(region, c)
    assert can_be_hoisted(region, d)
    assert not can_be_hoisted(region, e)
    assert not can_be_hoisted(region, f)
```

Reply:
I have address your comments